### PR TITLE
#89 executors now catch INT, TERM, QUIT in parent process

### PR DESCRIPTION
**Pull request by pwilczynskiclearcode**
#89 executors now catch INT, TERM, QUIT in parent process to kill subprocesses

### CODE REVIEW

**Comment by pwilczynskiclearcode** [comment]
test failing on pypy3:
```
=================================== FAILURES ===================================
___________________________ test_start_and_wait[10] ____________________________
timeout = 10
    @pytest.mark.parametrize('timeout', (None, 10))
    def test_start_and_wait(timeout):
        """Test if executor await for process to accept connections."""
        command = 'bash -c "sleep 2 && nc -l 3000"'
        executor = TCPExecutor(command, 'localhost', port=3000, timeout=timeout)
        executor.start()
    
>       assert executor.running() is True
E       assert <bound method TCPExecutor.running of <mirakuru.tcp.TCPExecutor: "bash -c "s..." 0x6773130>>() is True
E        +  where <bound method TCPExecutor.running of <mirakuru.tcp.TCPExecutor: "bash -c "s..." 0x6773130>> = <mirakuru.tcp.TCPExecutor: "bash -c "s..." 0x6773130>.running
command    = 'bash -c "sleep 2 && nc -l 3000"'
executor   = <mirakuru.tcp.TCPExecutor: "bash -c "s..." 0x6773130>
timeout    = 10
tests/executors/test_tcp_executor.py:22: AssertionError
```

File: src/mirakuru/base.py [hunk 1]
```diff
@@ -129,9 +130,26 @@ def __init__(
...
+        # We create a weak reference to allow garbage collection in case
+        # executor is deleted before the current process termination.
+        self_weak = weakref.ref(self)
+
+        def on_terminate(signum, frame):
+            executor = self_weak()
+            if executor:
+                executor.kill()
+
+        signal.signal(signal.SIGINT, on_terminate)
```
Comment (mmaslowskicc) [inline src/mirakuru/base.py]: The previous signal handler should be called too.

Comment (pwilczynskiclearcode) [inline src/mirakuru/base.py]: Which "previous" signal handler are you talking about?

Comment (mmaslowskicc) [inline src/mirakuru/base.py]: Any signal handler that might have been installed by the application using mirakuru, or by a different `SimpleExecutor` instance. https://docs.python.org/2.7/library/signal.html#signal.signal returns the previous handler if it's a Python function.

Comment (pwilczynskiclearcode) [inline src/mirakuru/base.py]: I will check if [atexit](https://docs.python.org/2/library/atexit.html) catches INT, TERM and QUIT signals then.

**Pull request closed without merging**